File: src/pgo/model/tla/TLAOperatorDefinition.java
package pgo.model.tla;

import java.util.List;
import java.util.stream.Collectors;

import pgo.util.SourceLocation;

/**
 * 
 * TLA AST PlusCalNode representing a defined operator.
 * 
 * op(...) == ...
 *
 */
public class TLAOperatorDefinition extends TLAUnit {
	private TLAIdentifier name;
	private List<TLAOpDecl> args;
	private TLAExpression body;
	private boolean local;
	
	public TLAOperatorDefinition(SourceLocation location, TLAIdentifier name, List<TLAOpDecl> args, TLAExpression body, boolean isLocal) {
		super(location);
		this.name = name;
		this.args = args;
		this.body = body;
		this.local = isLocal;
	}
	
	@Override
	public TLAOperatorDefinition copy() {
		return new TLAOperatorDefinition(getLocation(), name.copy(), args.stream().map(TLAOpDecl::copy).collect(Collectors.toList()), body.copy(), local);
	}
	
	@Override
	public <T, E extends Throwable> T accept(TLAUnitVisitor<T, E> v) throws E {
		return v.visit(this);
	}

	public TLAIdentifier getName() {
		return name;
	}

	public List<TLAOpDecl> getArgs() {
		return args;
	}

	public TLAExpression getBody() {
		return body;
	}
	
	public boolean isLocal() {
		return local;
	}

	@Override
	public int hashCode() {
		final int prime = 31;
		int result = 1;
		result = prime * result + ((args == null) ? 0 : args.hashCode());
		result = prime * result + ((body == null) ? 0 : body.hashCode());
		result = prime * result + (local ? 1231 : 1237);
		result = prime * result + ((name == null) ? 0 : name.hashCode());
		return result;
	}

	@Override
	public boolean equals(Object obj) {
		if (this == obj)
			return true;
		if (obj == null)
			return false;
		if (getClass() != obj.getClass())
			return false;
		TLAOperatorDefinition other = (TLAOperatorDefinition) obj;
		if (args == null) {
			if (other.args != null)
				return false;
		} else if (!args.equals(other.args))
			return false;
		if (body == null) {
			if (other.body != null)
				return false;
		} else if (!body.equals(other.body))
			return false;
		if (local != other.local)
			return false;
		if (name == null) {
			return other.name == null;
		} else return name.equals(other.name);
	}

}

File: src/pgo/model/tla/TLAOperatorDefinition.scala
package pgo.model.tla

import pgo.util.SourceLocation

import scala.jdk.CollectionConverters._

final class TLAOperatorDefinition(loc: SourceLocation, val name: TLAIdentifier, val args: List[TLAOpDecl],
                            val body: TLAExpression, val isLocal: Boolean) extends TLAUnit(loc) with TLADefinitionOne {
  def this(loc: SourceLocation, name: TLAIdentifier, args: java.util.List[TLAOpDecl], body: TLAExpression, isLocal: Boolean) =
    this(loc, name, args.asScala.toList, body, isLocal)

  override def copy: TLAOperatorDefinition =
    new TLAOperatorDefinition(loc, name.copy(), args.map(_.copy()), body.copy(), isLocal)

  override val definitions: List[TLADefinition] = List(this)

  def getName: TLAIdentifier = name
  def getArgs: java.util.List[TLAOpDecl] = args.asJava
  def getBody: TLAExpression = body

  override def arity: Int = args.length

  override def isModuleInstance: Boolean = false

  override def identifier: TLAIdentifier = name

  override def scope: Map[TLAIdentifier, TLADefinitionOne] = Map.empty

  override def accept[T, E <: Throwable](v: TLAUnitVisitor[T, E]): T = v.visit(this)

  override def equals(other: Any): Boolean = other match {
    case that: TLAOperatorDefinition =>
      name == that.name &&
        args == that.args &&
        body == that.body &&
        isLocal == that.isLocal
    case _ => false
  }

  override def hashCode(): Int = {
    val state = Seq(name, args, body, isLocal)
    state.map(_.hashCode()).foldLeft(0)((a, b) => 31 * a + b)
  }
}
